OneSignalSDK/onesignal/src/main/java/com/onesignal/OSSessionManager.java

```diff
@@ -88,112 +102,129 @@ Session fromString(String value) {
+            return;
+
+        OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG,
+           "OSSession changed" +
+              "\nfrom:\n" +
+              "session: " + this.session +
+              ", directNotificationId: " + this.directNotificationId +
+              ", indirectNotificationIds: " + this.indirectNotificationIds +
+              "\nto:\n" +
+              "session: " + session +
+              ", directNotificationId: " + directNotificationId +
+              ", indirectNotificationIds: " + indirectNotificationIds);
+
+        OutcomesUtils.cacheCurrentSession(session);
+        OutcomesUtils.cacheNotificationOpenId(directNotificationId);
+
+        // Broadcast the session that just ended before finalizing the state change
+        sessionListener.onSessionEnding(getSessionResult());
+
+        this.session = session;
+        this.directNotificationId = directNotificationId;
+        this.indirectNotificationIds = indirectNotificationIds;
     }
 
-    /**
-     * Set active session type to {@link Session#DIRECT}
-     */
-    void onSessionFromNotification(String notificationId) {
-        this.session = Session.DIRECT;
-        this.directSessionTime = System.currentTimeMillis();
-        this.notificationId = notificationId;
-        OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "Session Direct with notificationId: " + notificationId + "directSessionTime: " + directSessionTime);
+    private boolean willChangeSession(@NonNull Session session, @Nullable String directNotificationId, @Nullable JSONArray indirectNotificationIds) {
+        if (!session.equals(this.session))
+            return true;
+
+        // Allow updating a direct session to a new direct when a new notification is clicked
+        if (this.session.isDirect() &&
+                this.directNotificationId != null &&
+                !this.directNotificationId.equals(directNotificationId)) {
+            return true;
+        }
+
+        // Allow updating an indirect session to a new indirect when a new notification is received
+        if (this.session.isIndirect() &&
+           this.indirectNotificationIds != null &&
+           this.indirectNotificationIds.length() > 0 &&
+           !JSONUtils.compareJSONArrays(this.indirectNotificationIds, indirectNotificationIds)) {
+            return true;
+        }
+
+        return false;
     }
 
     /**
      * Set the notifications ids that influenced the session
      */
-    void setLastNotificationsId() {
+    @NonNull protected JSONArray getLastNotificationsReceivedIds() {
         JSONArray notificationsReceived = OutcomesUtils.getLastNotificationsReceivedData();
         JSONArray notificationsIds = new JSONArray();
 
-        long attributionWindow = OutcomesUtils.getIndirectAttributionWindow() * MIN_IN_MILLIS;
-        long currentTime = new Date().getTime();
+        long attributionWindow = OutcomesUtils.getIndirectAttributionWindow() * 60 * 1_000L;
```

Comment:
Maybe we should just create/use a helper like `convertTimeToMillis(Long time, TimeUnit fromUnits)`?